### PR TITLE
Add zero downtime migrations using elasticsearch aliases

### PR DESCRIPTION
**Pull request by fabianvf**
Purpose:
-----------
make ```inv migrate_search``` result in 0 search downtime, rather than completely deleting the search index and then reindexing from scratch.

Changes:
-------------
The search migration is now fully aware of elasticsearch (zero downtime migrations rely on fairly unique API features). The migration should be seamless, with all requests hitting a complete index.

Side-effects:
-----------------
This requires 2 copies of the index, so the memory footprint should be larger. Also, the original index is no longer deleted as an additional safety measure (so that breaking migrations can just quickly go back to the old index), although I can add that back in easily.

Notes:
----------
In case anyone is curious how it works but doesn't want to read the elasticsearch API docs:

* The first time this is run, I create an index called ```website_v1```
* I then copy the contents of the ```website``` index into ```website_v1```
* Then I delete ```website```, but I create an ```alias``` from ```website``` to ```website_v1``` (this means that all requests to ```localhost:9200/website``` are sent to ```localhost:9200/website_v1```)
* Subsequent times this is run, I create a new index, ```website_v{n+1}```, where ```n``` is the last index version
* I then go through the mongo database and add each user and node to the ```website_v{n+1}``` index
* When this is complete, I delete the old alias from ```website``` to ```website_v{n}``` and create a new alias from ```website``` to ```website_v{n+1}```
* Then, at our leisure, we can delete the old index, but in case we need it the data remains, and from the user's perspective, there was no gap in service.

### CODE REVIEW

**Comment by sloria** [comment]
Nice. @fabianvf Can you add tests for your new aliasing functions?

**Comment by fabianvf** [comment]
yes

File: tests/test_elastic.py [hunk 1]
```diff
@@ -478,3 +480,56 @@ def test_connection_error(self):
         )
         self.user.save()
         self.project.save()
+
+
+class TestSearchMigration(SearchTestCase):
+    """
```
Comment (sloria) [inline tests/test_elastic.py]: Prefer long names to docstrings for tests: http://cosdev.readthedocs.org/en/latest/process/testing.html#general-testing-guidelines

File: website/search_migration/migrate.py [hunk 2]
```diff
@@ -4,54 +4,107 @@
...
+    if delete:
+        delete_old(new_index)
 
     ctx.pop()
 
 
+def set_up_index(idx):
+    alias = es.indices.get_aliases(index=idx)
+
+    if not alias or not alias.keys() or idx in alias.keys():
```
Comment (sloria) [inline website/search_migration/migrate.py]: Nitpicky, but the check for `not alias.keys()` is probably unnecessary. If `alias` is a `dict` with no keys, then `not alias` will be falsy.

**Pull request merged**